File: output_repair/claude-opus/sample_4/repair_3/c/c/Struct/spouse.c
#include "stdlib.h"

struct person
{
  char *name;
  struct person *spouse;
};

/*@ predicate person(struct person *p;) =
      p->name |-> _ &*& p->spouse |-> _; @*/

//@ requires true;
//@ ensures person(result) &*& result->spouse |-> 0;
struct person *create_person()
{
  struct person *p = malloc(sizeof(struct person));
  if (p == 0)
    abort();
  p->spouse = 0;

  return p;
}

//@ requires p->name |-> _ &*& p->spouse |-> _ &*& q->name |-> _ &*& q->spouse |-> _;
//@ ensures p->name |-> _ &*& p->spouse |-> q &*& q->name |-> _ &*& q->spouse |-> p;
void marry(struct person *p, struct person *q)
{
  p->spouse = q;
  q->spouse = p;
}

//@ requires p->name |-> _ &*& p->spouse |-> ?s;
//@ ensures p->name |-> _ &*& p->spouse |-> s &*& result == s;
struct person *person_get_spouse(struct person *p)
{
  return p->spouse;
}

//@ requires p->name |-> _ &*& p->spouse |-> ?q &*& q != 0 &*& q->name |-> _ &*& q->spouse |-> p;
//@ ensures p->name |-> _ &*& p->spouse |-> 0 &*& q->name |-> _ &*& q->spouse |-> 0;
void divorce(struct person *p)
{
  p->spouse->spouse = 0;
  p->spouse = 0;
}

//@ requires person(p);
//@ ensures true;
void die(struct person *p)
{
  if (p->spouse != 0)
  {
    p->spouse->spouse = 0;
  }
  free(p);
}

//@ requires true;
//@ ensures true;
int main()
{
  struct person *alice = create_person();
  struct person *bob = create_person();
  struct person *eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}

File: output_20260318_091315/qwen3/java/java/Linklist/Spouse.java
/*@ predicate spouse_of(Person p, Person s) = p.spouse |-> s; @*/

class Person {

protected Person spouse;

//@ requires true;
//@ ensures true;
public void spouse_symm()
{
}

//@ requires true;
//@ ensures spouse_of(this, null);
public Person()
{
}

//@ requires spouse_of(this, ?s);
//@ ensures result == s;
public Person getSpouse()
{
    return spouse;
}

/*@ 
lemma void setSpouse_lemma(Person this, Person other)
    requires [?f1]spouse_of(this, ?s1) &*& [?f2]spouse_of(other, ?s2);
    ensures [f1]spouse_of(this, other) &*& [f2]spouse_of(other, this);
@*/
//@ requires [?f1]spouse_of(this, ?s1) &*& [?f2]spouse_of(other, ?s2);
//@ ensures [f1]spouse_of(this, other) &*& [f2]spouse_of(other, this);
protected void setSpouse(Person other)
{
    spouse = other;
    other.spouse = this;
}

/*@ 
lemma void clearSpouse_lemma(Person this)
    requires spouse_of(this, ?s) &*& spouse_of(s, this);
    ensures spouse_of(this, null) &*& spouse_of(s, null);
@*/
//@ requires spouse_of(this, ?s) &*& spouse_of(s, this);
//@ ensures spouse_of(this, null) &*& spouse_of(s, null);
protected void clearSpouse()
{
    spouse.spouse = null;
    spouse = null;
}

//@ requires [?f1]spouse_of(this, null) &*& [?f2]spouse_of(other, null);
//@ ensures [f1]spouse_of(this, other) &*& [f2]spouse_of(other, this);
void marry(Person other)
{
    other.setSpouse(this);
}

//@ requires spouse_of(this, ?s) &*& spouse_of(s, this);
//@ ensures spouse_of(this, null) &*& spouse_of(s, null);
void divorce()
{
    spouse.clearSpouse();
}

}

class Program {

//@ requires true;
//@ ensures true;
public static void main(String[] args)
{
    Person a = new Person();
    Person b = new Person();
    a.marry(b);
    b.divorce();
}

}